oak/server/buffer_channel.h

```diff
@@ -22,38 +22,45 @@
 
 namespace oak {
 
-// A channel implementation that forwards read and writes to underlying local buffers, keeping track
-// of read and write cursors.
-//
-// The channel owns the output buffer (std::vector), but does not own the input buffer (absl::Span).
-class BufferChannel final : public Channel {
+// A channel implementation that only has a receive half, which reads from a fixed
+// buffer provided at construction.  The channel does not own the provided buffer,
+// so the caller must ensure its lifetime is longer than that of the channel.
+class ReadBufferChannelHalf final : public ChannelHalf {
  public:
-  BufferChannel(absl::Span<const char> data_input, std::vector<char>* data_output)
-      : data_input_(data_input), data_output_(data_output){};
+  ReadBufferChannelHalf(absl::Span<const char> data) : data_(data) {}
 
   absl::Span<const char> Read(uint32_t size) override {
     LOG(INFO) << "Reading from channel: " << size << " bytes";
-    absl::Span<const char> data = data_input_.subspan(0, size);
-    data_input_.remove_prefix(data.size());
+    absl::Span<const char> data = data_.subspan(0, size);
+    data_.remove_prefix(data.size());
     return data;
-  };
+  }
+
+ private:
+  // This is a view of the data which advances each time the Read method is called.
+  absl::Span<const char> data_;
+};
+
+// A channel implementation that only has a send half, which writes to an output
+// buffer.  The channel does not own the provided buffer, so the caller must ensure
+// its lifetime is longer than that of the channel.
+class WriteBufferChannelHalf final : public ChannelHalf {
+ public:
+  WriteBufferChannelHalf(std::vector<char>* data) : data_(data) {}
 
   uint32_t Write(absl::Span<const char> data) override {
     LOG(INFO) << "Writing to channel: " << data.size() << " bytes";
-    if (data_output_ == nullptr) {
+    if (data_ == nullptr) {
```

Comment:
We should probably check for this in the constructor, now that the two halves are distinct there is no point in creating this half with a null data.

Reply:
Added a check in the constructor, but we can't do more than log an error there without more major surgery (e.g. replacing the constructor with a factory method).